### PR TITLE
RFC: Add Atomic{Ptr} support and tests

### LINKED ISSUE #29943: Support `Threads.Atomic{Ptr}`
There is a TODO for this in the code, but no open issue AFAICT.

https://github.com/JuliaLang/julia/blob/0fde275effe0103ba2aecf383e11c8d8fae229b1/base/atomics.jl#L27

That todo used to be for `Bool` as well, but that was solved in #26542

Without atomic operations on pointers, it is impossible to make lock-free multithreaded data structures.
which is a huge bummer for:
https://github.com/JuliaCollections/ThreadSafeDataStructures.jl

### PR DESCRIPTION
**Pull request by dillondaudert**
This defines the following atomic operations on pointers via `Atomic{Ptr}`:
```julia
get_index(::Atomic{Ptr})
set_index!(::Atomic{Ptr}, ::Ptr)
atomic_cas!(::Atomic{Ptr}, ::Ptr, ::Ptr)
atomic_xchg!(::Atomic{Ptr}, ::Ptr)
atomic_max!(::Atomic{Ptr}, ::Ptr}
atomic_min!(::Atomic{Ptr}, ::Ptr}
atomic_add!(::Atomic{Ptr}, ::Integer)
atomic_sub!(::Atomic{Ptr}, ::Integer)
```

Fixes #29943, and possibly addresses #32455.

This still needs NEWS.

### CODE REVIEW

File: base/atomics.jl
Comment (NHDaly) [inline base/atomics.jl]:
```suggestion
```

File: base/atomics.jl
Comment (NHDaly) [inline base/atomics.jl]: At first, I thought `Ptr` can't be added to `arithmetictypes` for the same reason that `Bool` couldn't: it doesn't support the same sets of operations. (You can't add two `Ptr`s together, etc). But in fact, I think it's fine since `arithmetictypes` is simply controlling _whether_ we'll define `atomic_add!` and `atomic_sub!`, which we do want to do. 👍 So we'll just need a separate `if` case there to define the `Ptr`-specific methods (See my comment below 🙂).

File: base/atomics.jl
Comment (NHDaly) [inline base/atomics.jl]: Please update this docstring to add `Ptr` to the list of supported types! :)

File: base/atomics.jl
Comment (NHDaly) [inline base/atomics.jl]: Here, you'll have to address the fact that `Ptr` doesn't support the same types of `add!` and `sub!` that Integers do:
```julia
julia> Ptr{Int}(0) + Ptr{Int}(1)
ERROR: MethodError: no method matching +(::Ptr{Int64}, ::Ptr{Int64})

julia> Ptr{Int}(0) - Ptr{Int}(1)
0xffffffffffffffff

julia> typeof(Ptr{Int}(0) - Ptr{Int}(1))
UInt64
```

It needs to be handled separately, so that you can add only `atomic_{add!,sub!}(Atomic{<:Ptr}, ::Int)`:
```
[2] +(x::Ptr, y::Integer) in Base at pointer.jl:159
[4] -(x::Ptr, y::Integer) in Base at pointer.jl:160
```

Likely this will have to just be a separate `if` case, rather than combining it with `Integer` as you've done here.
I think you'll want to define two methods: one generic one that converts all `<:Integer` types to `UInt` (as is done here): https://github.com/JuliaLang/julia/blob/74f2de13727373b44da76a65ceb5297889d3e482/base/pointer.jl#L159

And then define the `llvmcall` method for `atomic_add!(Atomic{<:Ptr}, ::UInt)`.

----------------------------------------------------

It is of course too bad that the actual `llvm` string will be the same in both `if`-cases. If you do want to reuse it, I guess you could consider moving it out into a helper variable before this for-loop block, perhaps?

File: test/threads_exec.jl
Comment (NHDaly) [inline test/threads_exec.jl]: Since you've added `Ptr` to the `arithmetictypes` above, you'll need to test `atomic_add!` and `atomic_sub!` here as well. :)

**Review by NHDaly — CHANGES_REQUESTED** [review]
This is awesome. Thanks for doing this, Dillon! :) Looking forward to it!

This will probably also need NEWS as well, I would think, but you may want to wait for someone else to confirm that.

File: base/atomics.jl
Comment (dillondaudert) [reply base/atomics.jl]: The available methods on `Ptr` are a little strange to me. Do you think it isn't worth having `atomic_{add!,sub!}` defined with `Ptr` as the second argument at all?

As for adding/subtracting `<:Integer`, the intrinsics `add_ptr` and `sub_ptr` call LLVM's `getelementptr` which performs pointer arithmetic, although it seems that it's always indexing [elements of type `i8`](https://github.com/JuliaLang/julia/blob/master/src/intrinsics.cpp#L1026). It's not clear to me why this is the case; without looking into it, I might have guessed the arithmetic would increment by `sizeof(T)` for a `Ptr{T}`. 

I wonder if the atomic version should also use `getelementptr` (and then presumably use `atomic_cas!`, similar to some of the current atomic float ops) or just call the `atomicrmw add/sub` instructions. Due to how `add_ptr` and `sub_ptr` are defined at the moment, I think these two methods would be the same only because pointer arithmetic always increments 1 byte at a time.

Comment (NHDaly) [reply base/atomics.jl]: > The available methods on `Ptr` are a little strange to me. Do you think it isn't worth having `atomic_{add!,sub!}` defined with `Ptr` as the second argument at all?

:) Yeah, I can see how it's unintuitive, but no, i think the methods provided by Base are correct, and we should follow them, here. The idea is basically that it is non-meaningful to add two pointers together.

As an analogy, I would say to think of Pointers are basically _points_ on a 1-dimensional line, and Integers as _vectors_ (or _offsets_) on that line. If you subtract two points, you get the vector between them. If you add a vector to a point, you get a new point. And you can add vectors to get a new vector. But it doesn't make sense to add two _points_ together. It's not really even clear what that means, because a point doesn't have a quantity; it's a dimensionless value.

More concretely, `Ptr`s are usually assigned by the kernel's malloc implementation, to some random position in memory. It is almost never going to be the case that adding together two such positions yields a valid position somewhere else in memory. Finally, as one more piece of evidence, for the reasons above, you can't add together pointers in C++ either:
```c++
input_line_6:2:4: error: invalid operands to binary expression ('int *' and 'int *')
 a + b
 ~ ^ ~
```
And similarly, in C++ subtracting two pointers returns a `long` (int).
Does that help clarify things? :) So yes, I think we should not define `atomic_add!(::Ptr, ::Ptr)` or `atomic_sub!(::Ptr, ::Ptr)`.

Comment (NHDaly) [reply base/atomics.jl]: > although it seems that it's always indexing [elements of type `i8`](https://github.com/JuliaLang/julia/blob/master/src/intrinsics.cpp#L1026). It's not clear to me why this is the case; without looking into it, I might have guessed the arithmetic would increment by `sizeof(T)` for a `Ptr{T}`.

Yeah, this surprised me as well when I came to Julia from C++. It's a different decision than is made in C++.
I'm not entirely sure why, but it's the way pointer arithmetic works in julia. If one wants to increment by whole objects, you use `p += sizeof(eltype(p))`. I _think_ the reason is that unlike in C, pointer arithmetic is rarely used in julia to iterate through an array, instead it seems to be more commonly used to offset into fields (bytes) within an element, so the decision was made to make that easier (rather than forcing you to `reinterpret` the pointer to be a `Ptr{Char}` just to index into it, as you would have to do in C/C++). That's just a guess on my part though! :)
Anyway, yeah, we'll want to support the same of course, where we're always indexing elements of `i8`.



> I wonder if the atomic version should also use `getelementptr` (and then presumably use `atomic_cas!`, similar to some of the current atomic float ops) or just call the `atomicrmw add/sub` instructions. Due to how `add_ptr` and `sub_ptr` are defined at the moment, I think these two methods would be the same only because pointer arithmetic currently always increments 1 byte at a time.

Ah, at first I was going to write that I agree with you that it doesn't matter, but no, it'll have to be implemented using `atomicrmw add/sub`. The reason is that the addition must be _atomic_, whereas `getelementptr` + `atomic_cas!` would require _first_ reading the value, then incrementing it, and _then_ assigning the value via `cas!`. But of course it may have changed in between those, which the `cas!` would then overwrite! So it must be `atomicrmw`, I think. :)

In which case, I think probably just sharing the `llvm` string with integer `add/sub` makes sense to me! :)

Comment (dillondaudert) [reply base/atomics.jl]: That all makes sense. I don't think contesting the choices for Julia's pointer arithmetic is really the point of this PR anyway, it just took me some time to figure out what was going on.

I'll go ahead and implement `atomic_{add!,sub!}(::Ptr, ::Integer)` using `atomicrmw {add,sub}`. It's equally possible to do so using `atomic_cas!`, but that's likely less efficient (see how this is done for floats [here](https://github.com/JuliaLang/julia/blob/master/base/atomics.jl#L421)) and not necessary in this case.

Comment (NHDaly) [reply base/atomics.jl]: > I'll go ahead and implement `atomic_{add!,sub!}(::Ptr, ::Integer)` using `atomicrmw {add,sub}`. It's equally possible to do so using `atomic_cas!`, but that's likely less efficient (see how this is done for floats [here](https://github.com/JuliaLang/julia/blob/master/base/atomics.jl#L421)) and not necessary in this case.

Oh, cool, makes sense. I hadn't seen that!
But yep, the `atomicrmw` plan sounds great. Lemme know when i should take a look again! :) thanks

File: base/atomics.jl
Comment (dillondaudert) [reply base/atomics.jl]: I actually removed it from `arithmetictypes` because even though `atomic_add!/sub!` exist for `Ptr`, they're different enough that it makes sense to me to not include it.

**Review by dillondaudert — COMMENTED** [review]
I defined `atomic_add!/sub!(::Atomic{Ptr}, ::Integer)` separately since the signature is different enough (everything else follows the pattern `(::Atomic{V}, V)`).

File: base/atomics.jl
Comment (NHDaly) [inline base/atomics.jl]: nit:
```suggestion
# Define atomic pointer +/- integer operations. Defined separately because they take `Int`s as the rhs.
```

File: base/atomics.jl
Comment (NHDaly) [inline base/atomics.jl]: @dillondaudert I've also only just now noticed that we're defining `and`, `nand`, `or, and `nor`, unfortunately none of which are defined for `Ptr` in base, as far as I can tell. :'(

It will be annoying to move those around, but maybe you want to move those into something like `logical_ops`, and not define them for Ptr, via the same logic as we use for the `arithmetic_ops`?
Sorry, that's annoying...

**Review by NHDaly — APPROVED** [review]
:) Thanks for doing this Dillon, it looks good to me!
(Other than the annoying thing I just noticed about the "logical ops". i've left a comment.)

I think this would be ready for review from a maintainer.

**Comment by dillondaudert** [comment]
I think this is ready for review. The build error looks unrelated.

**Comment by StefanKarpinski** [comment]
@vtjnash, are you the right person to review this?

**Comment by dillondaudert** [comment]
I can remove the logical ops defined on `Ptr` if reviewers think that's appropriate.

**Comment by dillondaudert** [comment]
I went ahead and excluded defining the logical operations on `Atomic{Ptr}`.

File: base/atomics.jl
Comment (dillondaudert) [reply base/atomics.jl]: I went ahead and implemented your suggestion.

**Review by NHDaly — APPROVED** [review]
Thanks, looks great!

**Comment by dillondaudert** [comment]
Bump

File: base/atomics.jl
Comment (JeffBezanson) [inline base/atomics.jl]: This should probably be `typ <: Integer || typ <: Ptr`. Otherwise it includes the type `Union{Integer, Ptr}`.

**Comment by JeffBezanson** [comment]
Welcome, @dillondaudert , and thanks for working on this. This looks basically correct, but I'm not sure what the use cases are. #29943 asks for it for implementing lock-free data structures, but I doubt it's actually useful for that since `Ptr`s are not tracked by the GC.

**Comment by dillondaudert** [comment]
Could you elaborate your thoughts on why not being tracked by the GC makes atomic `Ptr` not useful for lock-free data structures? I admit I don't know all the subtleties and consequences around that. 

I assumed since `Ptr` existed at all, it would be useful to have atomic operations for it. If that's not the case, should instead they be defined for `Ref`? I thought I read someone propose that in an issue at some point.

**Comment by rai-nhdaly** [comment]
I could maybe imagine this being useful for something like Blobs.jl, making
use of the atomic add/subtract. But yeah, otherwise simple pointer
assignment is of course already atomic. Maybe we would use compare and swap
in something like Blobs as well?

I admit I haven't thought about it too much.

Definitely agree with Jeff's assessment that this alone won't be terribly
useful for lock-free datastructures. It would probably be most useful to
just support arbitrary types, with separate support for stack vs heap
allocated types?

**Comment by oxinabox** [comment]
> But yeah, otherwise simple pointer assignment is of course already atomic.

No, it isn't.
 I don't have a source handy but I found one for when @iamed2 made that claim. I can dig though LibPQ PRs and find it again tomorrow.
Pointer assignment is not atomic.
The CPU can act mid-word.


---

Yes, we likely also want this for `Ref`.
The TODO note in the source code only requested for `Ptr`.
Also seemed like that would be easier from working out what Todo with llvm.

**Comment by vtjnash** [comment]
> But yeah, otherwise simple pointer assignment is of course already atomic.

It's unclear exactly what you intended to mean here. The simple assignment isn't "atomic" because that's not a property of the assignment but of the serializing side-effects that occur simultaneously (in the compiler and/or hardware).  OTOH, it's pretty much unheard of for a processor do have a cache line smaller than a pointer word size, so the write of a pointer-sized word to an address with valid alignment is usually free of data races (torn writes). It's important to distinguish though that the Julia `unsafe_load` intrinsics though do not require alignment, and thus are NOT safe to use for threaded read/writes (the similar operation on a normal Julia object and array references should be fine).

> I assumed since Ptr existed at all, it would be useful to have atomic operations for it

Yes, but maybe worth clarifying that's not what this PR does. It's been my goal for some time now to provide new `Core.Intrinsics` for doing atomic versions of each of the existing ops (e.g. fieldref, arrayref, pointerref, etc.) and then deprecate this file (it's inefficient and awkward). I'm happy to have others contribute towards the work to make it happen sooner, though I'd encourage anyone thinking of working in that direction to make a Julep first so we can review the design before they put a lot of effort into coding something.

**Comment by dillondaudert** [comment]
I'd be happy to help get proper atomic (pointer) operations implemented in Julia. At this point, I don't believe I'm familiar enough with `Core.Intrinsics` (or the internals in general) well enough to make a design proposal. Since the changes in this PR aren't the right way to go about this, I'll close it.

**Comment by oxinabox** [comment]
This remains an operation we should have though?

**Pull request closed without merging**
